cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java

```diff
@@ -85,76 +91,178 @@ public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runna
   @POST
   @Path("/metadata")
   public void metadata(FullHttpRequest request, HttpResponder responder) throws Exception {
-    String requestBody = request.content().toString(StandardCharsets.UTF_8);
+    Map<String, GenericRecord> consumeRequests = decodeConsumeRequest(request);
+    MessagesBodyProducer messagesBodyProducer = new MessagesBodyProducer(fetchAllMessages(consumeRequests),
+                                                                         consumeRequests.size());
+    responder.sendContent(HttpResponseStatus.OK, messagesBodyProducer,
+                          new DefaultHttpHeaders().set(HttpHeaderNames.CONTENT_TYPE, "avro/binary"));
+  }
+
+  /**
+   * shuts down remote runtime
+   */
+  @POST
+  @Path("/shutdown")
+  public void shutdown(FullHttpRequest request, HttpResponder responder) throws Exception {
+    responder.sendString(HttpResponseStatus.OK, "Triggering shutdown down Runtime Http Server.");
+    shutdownRunnable.run();
+  }
+
+  /**
+   * Decode consume request from avro binary format
+   */
+  private Map<String, GenericRecord> decodeConsumeRequest(FullHttpRequest request) throws IOException {
+    Decoder decoder = DecoderFactory.get().directBinaryDecoder(new ByteBufInputStream(request.content()), null);
+    Schema requestSchema = MonitorSchemas.V1.MonitorConsumeRequest.SCHEMA;
+    // Avro converts java String to UTF8 while deserializing strings. Set String type to Java String
+    GenericData.setStringType(requestSchema, GenericData.StringType.String);
+    DatumReader<Map<String, GenericRecord>> datumReader = new GenericDatumReader<>(requestSchema);
+
+    return datumReader.read(null, decoder);
+  }
+
+  /**
+   * Fetches all the messages for each topic provided in consumeRequests and returns iterator of topicConfig
+   * to message iterator
+   */
+  private Iterator<Map.Entry<String, CloseableIterator<Message>>> fetchAllMessages(Map<String, GenericRecord>
```

Comment:
Better not to open all iterators for all topics. Why not open it one by one since you are responding one by one anyway?